cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,298 @@
+		return nil
+	}
+
+	// Check if the directory is empty.
+	ipamDir, err := os.Open(ipAllocPath)
+	if err != nil {
+		return fmt.Errorf("failed to open k8s-pod-network dir: %s", err)
+	}
+	if _, err = ipamDir.Readdirnames(1); err != nil {
+		if err == io.EOF {
+			log.Info("k8s-pod-network dir empty; no migration necessary, successfully exiting...")
+			return nil
+		}
+		return fmt.Errorf("failed to read k8s-pod-network dir names: %s", err)
+	}
+	defer func() {
+		if err = ipamDir.Close(); err != nil {
+			log.WithError(err).Errorf("failed to close k8s-pod-network directory")
+		}
+	}()
+
+	// Disable cni by setting datastoreReady to false.
+	clusterInfo, err := c.ClusterInformation().Get(ctxt, "default", options.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to fetch cluster information: %s", err)
+	}
+	if *clusterInfo.Spec.DatastoreReady {
+		*clusterInfo.Spec.DatastoreReady = false
+		if clusterInfo, err = c.ClusterInformation().Update(ctxt, clusterInfo, options.SetOptions{}); err != nil {
+			return fmt.Errorf("failed to disable cluster: %s", err)
+		}
+	}
+
+	// Also disable cni by disabling the binaries.
+	for _, binary := range binariesToDisable {
+		dir, file := path.Split(binary)
+		disabledBinary := path.Join(dir, file+".disabled")
+		if err = os.Rename(binary, disabledBinary); err != nil {
+			return fmt.Errorf("failed to disable binary %s: %s", binary, err)
+		}
+	}
+
+	podList, err := k8sClient.CoreV1().Pods("").List(metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("spec.nodeName=%s", nodename),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to list pods: %s", err)
+	}
+	podIPMap := make(map[string]*corev1.Pod)
+	for i := 0; i < len(podList.Items); i++ {
+		podIPMap[podList.Items[i].Status.PodIP] = &podList.Items[i]
+	}
+
+	// Read in all the files in the host-local directory.
+	files, err := ioutil.ReadDir(ipAllocPath)
+	if err != nil {
+		return fmt.Errorf("failed to read path %s: %s", ipAllocPath, err)
+	}
+
+	// For each file, convert it into an IP allocation and then delete the file.
+	for _, f := range files {
+		fname := path.Join(ipAllocPath, f.Name())
+
+		// Delete and skip the last reserved IP.
+		if strings.Contains(f.Name(), "last") || f.Name() == "lock" {
+			if err = os.Remove(fname); err != nil {
+				return fmt.Errorf("failed to remove file %s: %s", fname, err)
+			}
+			continue
+		}
+
+		// The name of the file is its IP address.
+		ip, _, err := cnet.ParseCIDR(fmt.Sprintf("%s/32", f.Name()))
+		if err != nil {
+			return fmt.Errorf("failed to parse IP %s: %s", f.Name(), err)
+		}
+
+		// The contents are the container ID.
+		b, err := ioutil.ReadFile(fname)
+		if err != nil {
+			return fmt.Errorf("failed to read file %s: %s", fname, err)
+		}
+		containerID := string(b)
+
+		// See if the IP hasn't already been assigned for the handleID.
+		handleID := utils.GetHandleID("k8s-pod-network", containerID)
+		/*
+			ips, err := c.IPAM().IPsByHandle(ctxt, handleID)
```

Comment:
What's this for? Can we delete it?

Reply:
Forgot to finish implementing the logic to check if the IP hasn't already been assigned for the given handle